pkg/docker/docker.go

```diff
@@ -0,0 +1,310 @@
+	// the container
+	CgroupMemorySize size.ByteCount
+
+	// RestartPolicy_ determines what behavior docker should have when the pod exits.
+	// Typically it will indicate that docker should always restart the container
+	// for long-running services (e.g. servers)
+	RestartPolicy_ runit.RestartPolicy
+}
+
+func (l *Launchable) Disable() error {
+	// TODO: implement
+	return nil
+}
+
+func (l *Launchable) EnvDir() string {
+	return filepath.Join(l.RootDir, "env")
+}
+
+func (l *Launchable) EnvVars() map[string]string {
+	// TODO: implement
+	return nil
+}
+
+func (l *Launchable) Executables(_ *runit.ServiceBuilder) ([]launch.Executable, error) {
+	// We don't make use of runit for docker launchables, so there are no
+	// executables to supply. Instead we'll just send "docker run" and
+	// "docker stop" commands to dockerd.
+	return nil, nil
+}
+
+func (l *Launchable) GetRestartTimeout() time.Duration {
+	return l.RestartTimeout
+}
+
+func (l *Launchable) ID() launch.LaunchableID {
+	return l.LaunchableID
+}
+
+func (l *Launchable) InstallDir() string {
+	// There is no install directory for a docker container
+	return ""
+}
+
+func (l *Launchable) Installed() bool {
+	// docker pull is idempotent and cheap, so don't worry about this
+	return false
+}
+
+func (l *Launchable) Launch(_ *runit.ServiceBuilder, _ runit.SV) error {
+	if l.DockerClient == nil {
+		return util.Errorf("docker client was not initialized, can't launch docker launchable")
+	}
+
+	ctx := context.TODO()
+
+	// get pod env vars from disk
+	podEnvVars, err := loadEnvVars(l.PodEnvDir)
+	if err != nil {
+		return err
+	}
+	// get launchable env vars from disk
+	launchableEnvVars, err := loadEnvVars(l.EnvDir())
+	if err != nil {
+		return err
+	}
+	envVars := append(podEnvVars, launchableEnvVars...)
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return util.Errorf("could not get hostname: %v", hostname)
+	}
+	containerConfig := &container.Config{
+		Hostname: hostname,
+		User:     l.RunAs,
+		Env:      envVars,
+		Image:    l.Image,
+	}
+
+	restartPolicy := "always"
+	switch l.RestartPolicy_ {
+	case runit.RestartPolicyAlways:
+		// keep it the way it is
+	case runit.RestartPolicyNever:
+		restartPolicy = "no"
+	default:
+		return util.Errorf("invalid restart policy: %s", l.RestartPolicy_)
+	}
+
+	bindings := []string{}
+	dockerMountsDir := filepath.Join(l.PodHomeDir, "docker_mounts.d")
+	mountFiles, err := ioutil.ReadDir(dockerMountsDir)
+	if err != nil && !os.IsNotExist(err) {
+		return util.Errorf("could not read from %s dir: %v", dockerMountsDir, err)
+	}
+	for _, file := range mountFiles {
+		mounts := Mounts{}
+		if !file.IsDir() {
+			// read content and load into bindings var
+			mountFilePath := filepath.Join(dockerMountsDir, file.Name())
+			data, err := ioutil.ReadFile(mountFilePath)
+			if err != nil {
+				return util.Errorf("could not read file %s: %v", mountFilePath, err)
+			}
+			err = yaml.Unmarshal(data, mounts)
```

Comment:
don't you need `&mounts`? I'd be surprised if it gets filled in if this isn't a pointer

Reply:
Thanks for catching that!

Reply:
I recall it did work before during testing oddly but this is the correct way.